opentelemetry-sdk/src/metrics/instrument.rs

```diff
@@ -218,52 +217,57 @@ impl Stream {
     }
 }
 
-/// the identifying properties of a stream.
+/// The identifying properties of a instrument.
 #[derive(Debug, PartialEq, Eq, Hash)]
-pub(crate) struct StreamId {
+pub(crate) struct InstId {
     /// The human-readable identifier of the stream.
     pub(crate) name: Cow<'static, str>,
     /// Describes the purpose of the data.
     pub(crate) description: Cow<'static, str>,
+    /// Defines the functional group of the instrument.
+    pub(crate) kind: InstrumentKind,
     /// the unit of measurement recorded.
     pub(crate) unit: Unit,
-    /// The stream uses for an instrument.
-    pub(crate) aggregation: String,
-    /// Monotonic is the monotonicity of an instruments data type. This field is
-    /// not used for all data types, so a zero value needs to be understood in the
-    /// context of Aggregation.
-    pub(crate) monotonic: bool,
-    /// Temporality is the temporality of a stream's data type. This field is
-    /// not used by some data types.
-    pub(crate) temporality: Option<Temporality>,
     /// Number is the number type of the stream.
     pub(crate) number: Cow<'static, str>,
 }
 
+impl InstId {
+    /// Instrument names are considered case-insensitive.
+    ///
+    /// Standardize the instrument name to always be lowercase so it can be compared
+    /// via hash.
+    pub(crate) fn normalize(&mut self) {
+        if self.name.chars().any(|c| c.is_ascii_uppercase()) {
+            self.name = self.name.to_lowercase().into();
+        }
```

Comment:
Curious why `is_ascii_uppercase` was chosen to check for need to call `to_lowercase` as opposed to [`is_uppercase`](https://doc.rust-lang.org/std/primitive.char.html#method.is_uppercase)?

It occurs to me there could be a strange behavior where a name that contains both ASCII uppercase and other Unicode uppercase characters gets all lower-cased, but a name that only contains Unicode uppercase characters does not...

Reply:
The names are always ASCII so this is slightly more effecient. Made the lower conversion ascii too and added a reference to https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/metrics/api.md#instrument-name-syntax